### PR TITLE
Fix host_indices use-after-free segfault on prefetch abort

### PR DESCRIPTION
**Pull request by flymysql**
## Motivation

Under HiCache L3 storage prefetch with the `timeout`/`best_effort` policy, the scheduler can abort an in-flight prefetch while the storage IO thread is still reading `host_indices`. This causes a hard segfault (`Fatal Python error: Segmentation fault`) in the storage backend, which then cascades into `Connection closed by peer` / broken distributed groups on other TP ranks.

Root cause: `HostKVCache` subclass `alloc()` returns `self.free_slots[:need_size]`, a **view** into the pool's `free_slots` arange-backed storage. The allocated `host_indices` therefore aliases the pool's free list. On prefetch abort the scheduler thread runs `free()` (`torch.cat`) / a subsequent `alloc()` and rewrites `free_slots`, while the storage IO thread concurrently reads the same shared storage (e.g. `.tolist()` in the backend) — racing on storage/refcount → segfault.

## Modifications

- `memory_pool_host.py`: `alloc()` now `.clone()`s the selected indices in all three host pools (`MambaPoolHost`, `LogicalHostPool`, `DeepSeekV4PagedHostPool`), so returned `host_indices` own their storage and their lifetime is decoupled from pool structural mutation. Index tensors are tiny, so the copy cost is negligible.
- `cache_controller.py`: `prefetch_io_aux_func` skips `_page_transfer` for operations already terminated while queued in `prefetch_buffer` (abort-before-dispatch), avoiding backend IO into pages the scheduler is about to reclaim/reassign.

## Accuracy Test

- No behavior change on the happy path; `alloc` still returns the same index values.
- Fixes a reproducible segfault seen with `--hicache-storage-prefetch-policy timeout` on multi-pool (UnifiedRadixCache) DeepSeek-V4 / SWA setups under disk pressure.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Provide a link to related issues or PRs if applicable.

---
### CI States

Latest PR Test (Base): :x: [Run #29399448962](https://github.com/sgl-project/sglang/actions/runs/29399448962)
Latest PR Test (Extra): :x: [Run #29399448705](https://github.com/sgl-project/sglang/actions/runs/29399448705)

### CODE REVIEW

**Comment by flymysql** [comment]
### Follow-up: frame this without a proprietary backend

The previous comment mentioned an internal L3 storage backend path; please ignore that naming. The bug is entirely in SGLang's host-pool / HiCache prefetch lifecycle and is **backend-agnostic**.

**Trigger (public APIs only):**
- HiCache storage enabled (any `hicache-storage-backend` that uses zero-copy page get with `host_indices`)
- `UnifiedRadixCache` / multi-pool path (e.g. DeepSeek-V4)
- `--hicache-storage-prefetch-policy timeout` (default) or `best_effort`
- Prefetch slower than the timeout → scheduler aborts while the prefetch IO aux thread still reads `operation.host_indices`

**Failure mode:**
```text
Fatal Python error: Segmentation fault
```
in the storage IO thread while converting / indexing `host_indices` (any backend that touches that tensor after dispatch). Sibling TP ranks then die with gloo `Connection closed by peer`.

**Root cause (in-tree):**
`HostKVCache` subclass `alloc()` returned `self.free_slots[:need_size]` — a **view** into the free-list storage. After abort, the scheduler thread `free()`s / re-`alloc()`s (rewriting `free_slots`) while the IO thread still reads the shared storage → use-after-free / storage race.

**This PR:**
1. `.clone()` in the three host-pool `alloc()` paths so returned `host_indices` own storage.
2. Skip `_page_transfer` if the op was already terminated while queued in `prefetch_buffer` (abort-before-dispatch).

Happy to squash / rebase onto latest `main` if preferred.

**Comment by hzh0425** [comment]
I don’t think cloning host_indices fully fixes the underlying race. The allocator updates free_slots through slicing and torch.cat, which do not mutate the storage retained by an existing tensor view？

**Comment by hzh0425** [comment]
The root cause of this issue is that the extra pool lacks lifecycle management similar to that of KV, leading to conflicts.
This isn't a fundamental fix—I'll address it properly.

**Comment by flymysql** [comment]
Thanks @hzh0425 for the review — agreed on both points.

1. **Clone is not a fundamental fix**: `free_slots` updates via slice reassignment / `torch.cat` do not mutate the storage retained by an existing view; the index values remain readable. Our original hypothesis overstated the UAF risk at the tensor-storage level.

2. **Root cause is extra-pool lifecycle in UnifiedRadixCache**: the production crash was on the multi-pool (DeepSeek-V4 / SWA) path under prefetch timeout. The gap is likely in synchronizing extra-pool `PoolTransfer` host_indices alloc/free with KV on abort, not in generic `alloc()` returning a view.

Closing this PR and waiting for your proper fix. Happy to help validate on our side (HiCache L3 + UnifiedRadixCache + `--hicache-storage-prefetch-policy timeout`) once a patch is ready.

**Comment by flymysql** [comment]
Tracking the underlying fix in #31445.
